experimental/dds/tree2/src/class-tree/README.md

```diff
@@ -0,0 +1,49 @@
+# class-tree
+
+This is a temporary directory containing a second implementation of the tree types from simple-tree, but built on-top of a customized schema system optimized for this specific tree API.
+
+The main differences from the old schema system (SchemaBuilder and its related types):
+
+1. The new schema system is layered on-top of internal one without leaking all the types from it into the public API.
+   Wrappers, type erasure and casting in the implementation are used as required to accomplish this.
+2. The new system is designed from scratch to work at the same abstraction layer and data model as used in simple-tree.
+   This means concepts like `List` are supported in a first class way, and the focus is on kinds of nodes (List, Object, Map), and not on FieldKinds.
+   No extra special cases and carve-outs are needed to support cases the internal data-model supports:
+   tests for those cases will simple not use this wrapper layer and can use the pre-existing schema builder instead.
+3. Schema are classes, and the instance types of those classes are what gets exposed in the tree API.
+   This has many benefits:
+    1. Users don't have to use `typeof` or invoke any type meta-functions to get the node types they want to pass around: just use the class/schema name as the type.
+       This replaces `Typed<typeof myNodeSchema>` with just `MyNode`.
+    2. Recursive schema work in `d.ts`` files due to use of classes.
+       See [Generated d.ts includes implicit any for recursive types](microsoft/TypeScript#55832).
+    3. Intellisense is much cleaner when referring to types defined in schema:
+       it just uses the class name (for example `MyNode`) when referring to the node type or `typeof MyNode` when referring to the schema's type.
+       These simplifications are what resolve the `d.ts`` issue noted above with recursive types.
+    4. Normal JS/TS type narrowing with `instanceof` can be used with schema defined types.
+    5. It's possible to add view/session local state to instances as properties, as well as adding methods by just putting them in the class like any other class.
+
+Currently this implementation exposes the prototypes from these classes: this impacts some generic object based code.
+For example, this probably impacts Node's deep equals as follows:
+
+1. Two trees with the same "shape" but different types no longer compare equal.
+2. Trees no longer compare equal to plain objects with no prototypes.
+
+## Limitations
+
+Fields allowing `Any` can not be supported due to the requirement to traverse all types from the root.
+
+Returned classes from factory cannot have any private or protected members due to a [TypeScript limitation](https://github.com/microsoft/TypeScript/issues/36060).
+This means getting nominal typing (non-structural typing) of node will require explicit members (like a strongly typed schema or type name symbol) if nominal typing is desired.
```

Comment:
Our schema system is nominal. You can't use a Foo node as a Bar node, even if all of Bar's members are optional. Our typing does not reflect this (and we also lack runtime errors, so it can result in out of schema documents!). I think this is quite bad, and the exact same issue caused bugs in Whiteboard and we fixed it there, but I'm not going to personally state that this is something we should or will be doing as that beyond my authority here.